mvrx/src/main/kotlin/com/airbnb/mvrx/BaseMavericksViewModel.kt

```diff
@@ -164,6 +169,84 @@ abstract class BaseMavericksViewModel<S : MvRxState>(
         stateStore.get(action)
     }
 
+    /**
+     * Run a coroutine and wrap its progression with [Async] property reduced to the global state.
+     *
+     * @param dispatcher The coroutine dispatcher that the coroutine will run on. Defaults to [Dispatchers.Main.immediate].
+     * @param retainValue A state property that, when set, will be called to retrieve an optional existing data value that will be retained across
+     *                    subsequent Loading and Fail states. This is useful if you want to display the previously succcessful data when
+     *                    refreshing.
+     * @param reducer A reducer that is applied to the current state and should return the new state. Because the state is the receiver
+     *                and it likely a data class, an implementation may look like: `{ copy(response = it) }`.
+     */
+    fun <T : Any?> Deferred<T>.execute(
+        dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
+        retainValue: KProperty1<S, Async<T>>? = null,
+        reducer: S.(Async<T>) -> S
+    ) = suspend { await() }.execute(dispatcher, retainValue, reducer)
+
+    /**
+     * Run a coroutine and wrap its progression with [Async] property reduced to the global state.
+     *
+     * @param dispatcher The coroutine dispatcher that the coroutine will run on. Defaults to [Dispatchers.Main.immediate].
+     * @param retainValue A state property that, when set, will be called to retrieve an optional existing data value that will be retained across
+     *                    subsequent Loading and Fail states. This is useful if you want to display the previously succcessful data when
+     *                    refreshing.
+     * @param reducer A reducer that is applied to the current state and should return the new state. Because the state is the receiver
+     *                and it likely a data class, an implementation may look like: `{ copy(response = it) }`.
+     */
+    fun <T : Any?> (suspend () -> T).execute(
+        dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
+        retainValue: KProperty1<S, Async<T>>? = null,
+        reducer: S.(Async<T>) -> S
+    ): Job {
+        setState { reducer(Loading(value = retainValue?.get(this)?.invoke())) }
+        return viewModelScope.launch(dispatcher) {
+            try {
+                val result = invoke()
+                setState { reducer(Success(result)) }
+            } catch (e: CancellationException) {
+                @Suppress("RethrowCaughtException")
+                throw e
```

Comment:
why catch this if it's just going to be rethrown.

Also, should we have a discussion about expected behavior on cancellation? with this, if the job is cancelled the state will be left Loading indefinitely, which seems misleading. we could provide a way to handle this more clearly, like a Async.Canceled or we could set the property back to Uninitialized.

or maybe it makes sense to represent it as Fail? or make a Cancel object that is a subclass of Fail?